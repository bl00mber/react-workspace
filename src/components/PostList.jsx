import React from 'react';
import PropTypes from 'prop-types';

const PostList = props => (
  <div>
    {props.isFetching && <p>Fetching...</p>}
    {!props.isFetching && props.error && <p>Error: {props.error.description}</p>}
    <ul>
      {
        props.posts.map(post =>
          <li key={post.id}>{post.description}</li>)
      }
    </ul>
    <button onClick={() => props.onFetchPosts(true)}>Fetch Posts OK</button>
    <button onClick={() => props.onFetchPosts(false)}>Fetch Posts Errors</button>
  </div>
);

PostList.defaultProps = {
  error: null,
};

PostList.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  error: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  onFetchPosts: PropTypes.func.isRequired,
};

export default PostList;
