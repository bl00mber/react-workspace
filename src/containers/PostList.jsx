import { connect } from 'react-redux';

import PostListComponent from '@/components/PostList';
import { fetchPostsAsync } from '@/store/actions';

const mapStateToProps = state => ({
  isFetching: state.posts.isFetching,
  error: state.posts.error,
  posts: state.posts.posts,
});

const mapDispatchToProps = dispatch => ({
  onFetchPosts: params => fetchPostsAsync(params)(dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(PostListComponent);
