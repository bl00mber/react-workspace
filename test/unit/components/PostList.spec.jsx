import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import PostList from '@/components/PostList';

describe('PostList', () => {
  test('should match enzyme shallow snapshot', () => {
    const props = {
      isFetching: false,
      posts: [],
      onFetchPosts: jest.fn(),
    };
    const wrapper = shallow(<PostList {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setProps({
      isFetching: true,
    });
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setProps({
      isFetching: false,
      posts: [
        { id: 1, description: 'post 1' },
      ],
    });
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setProps({
      isFetching: true,
    });
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setProps({
      isFetching: false,
      posts: [],
      error: { description: 'fetching error' },
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('clicking on buttons should call onFetchPosts prop', () => {
    const props = {
      isFetching: false,
      posts: [],
      onFetchPosts: jest.fn(),
    };
    const wrapper = shallow(<PostList {...props} />);
    wrapper.find('button').at(0).simulate('click');
    expect(props.onFetchPosts).toHaveBeenCalledWith(true);
    props.onFetchPosts.mockReset();
    wrapper.find('button').at(1).simulate('click');
    expect(props.onFetchPosts).toHaveBeenCalledWith(false);
  });
});
