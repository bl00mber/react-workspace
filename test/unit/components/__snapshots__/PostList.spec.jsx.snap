// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList should match enzyme shallow snapshot 1`] = `
<div>
  <ul />
  <button
    onClick={[Function]}
  >
    Fetch Posts OK
  </button>
  <button
    onClick={[Function]}
  >
    Fetch Posts Errors
  </button>
</div>
`;

exports[`PostList should match enzyme shallow snapshot 2`] = `
<div>
  <p>
    Fetching...
  </p>
  <ul />
  <button
    onClick={[Function]}
  >
    Fetch Posts OK
  </button>
  <button
    onClick={[Function]}
  >
    Fetch Posts Errors
  </button>
</div>
`;

exports[`PostList should match enzyme shallow snapshot 3`] = `
<div>
  <ul>
    <li
      key="1"
    >
      post 1
    </li>
  </ul>
  <button
    onClick={[Function]}
  >
    Fetch Posts OK
  </button>
  <button
    onClick={[Function]}
  >
    Fetch Posts Errors
  </button>
</div>
`;

exports[`PostList should match enzyme shallow snapshot 4`] = `
<div>
  <p>
    Fetching...
  </p>
  <ul>
    <li
      key="1"
    >
      post 1
    </li>
  </ul>
  <button
    onClick={[Function]}
  >
    Fetch Posts OK
  </button>
  <button
    onClick={[Function]}
  >
    Fetch Posts Errors
  </button>
</div>
`;

exports[`PostList should match enzyme shallow snapshot 5`] = `
<div>
  <p>
    Error: 
    fetching error
  </p>
  <ul />
  <button
    onClick={[Function]}
  >
    Fetch Posts OK
  </button>
  <button
    onClick={[Function]}
  >
    Fetch Posts Errors
  </button>
</div>
`;
